### PR TITLE
feat: Allow DistributedSampler to accept Sampler as input

### LINKED ISSUE #23430: [Feature request] Let DistributedSampler take a Sampler as input
## Motivation

Currently, `DistributedSampler` assumes that it takes a `Dataset` as argument. But in reality, the only information it exploits from it is its `len`.

We sometimes want to have a custom Sampler to be used in distributed mode. So it might be desirable to also let `DistributedSampler` take a `Sampler` as argument.

## Potential implementation

The only difference is that in 

https://github.com/pytorch/pytorch/blob/46224ef89edcd0e350d1572bb4d12d370245cc6e/torch/utils/data/distributed.py#L57-L61
We would additionally have something like

```python
if isinstance(self.dataset, Sampler):
    orig_indices = list(iter(self.dataset))
    indices = [orig_indices[i] for i in indices]

return iter(indices)
```

## Pitch

More modularity and code reuse
```python
sampler = MyNiceSampler(dataset)
if distributed:
    sampler = DistributedSampler(sampler)
```

## Alternatives

We can integrate the functionality of `DistributedSampler` inside our custom sampler, but this seems redundant.

### PR DESCRIPTION
**Pull request by ankitlade12**
Fixes #23430

This change enables composition of sampling strategies in distributed training by allowing DistributedSampler to wrap another Sampler.

Changes:
- Add Union[Dataset, Sampler] type hint to __init__
- Add _input_is_sampler flag for runtime detection
- Update __iter__ to use wrapped sampler's indices when applicable
- Update docstring with new usage examples
- Add 7 comprehensive unit tests covering various sampler types

Example usage:
    base_sampler = WeightedRandomSampler(weights, num_samples)
    sampler = DistributedSampler(base_sampler, shuffle=False)

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Some design concerns from the original issue:
- the wrapper could invoke the child sampler's `set_epoch` if it exists (or the method could be named differently, e.g. `set_seed`): https://github.com/pytorch/pytorch/issues/23430#issuecomment-1038032217
- there could be two different ways to split the data by replicas: contiguous chunks or interleaved chunks (especially when the child sampler is non-shuffling). E.g. if the dataset is sorted by seqlen, this choice can affect whether total seqlen varies across replicas and affect GPU utilization: https://github.com/pytorch/pytorch/issues/23430#issuecomment-1038033413 https://github.com/pytorch/pytorch/issues/23430#issuecomment-940457568

**Comment by ankitlade12** [comment]
I've addressed both concerns in this implementation:

1. set_epoch propagation: I've updated the code so that calling DistributedSampler.set_epoch will now automatically call .set_epoch() (or .set_seed()) on the wrapped sampler if those methods exist. This keeps the child sampler in sync.

2. Splitting strategies: I added a new  split_contiguous parameter.

- By default, it uses interleaved chunks (same as before).
- If set to True, it uses contiguous chunks. This solves the issue where sorted datasets could cause unbalanced workloads across GPUs.

I've added 10 tests to verify that both the epoch propagation and the splitting strategies work as expected.

**Comment by ankitlade12** [comment]
Hey @divyanshk @ramanishsingh @aelavender — just wanted to bump this PR. It was recently marked as Stale, but it's ready for review. I've addressed the design concerns raised by @vadimkantorov (set_epoch propagation and contiguous vs. interleaved splitting) and added tests for both.

Would really appreciate it if you could take a look when you get a chance. Happy to make any changes based on your feedback. Thanks!
